Blog: The self-driving product (and why we're launching PostHog Code) (#16500)
PostHog/posthog.com: contents/blog/self-driving-product.mdx

---
date: "2026-04-20"
title: "The self-driving product (and why we're launching PostHog Code)"
author:
  - cleo-lant
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/selfdriving_3f4c644a54.png
featuredImageType: full
category: PostHog news
tags:
  - Product updates
  - PostHog news
---

import { WaitlistForm } from 'components/WaitlistForm'

Today, [PostHog Code](/code) enters beta. It's a desktop app that runs coding agents on top of your product data. The obvious stuff ships itself. The tricky stuff shows up as a prioritized to-do list for you to steer.

> We built it to chase one idea: the self-driving product. A product that spots its own problems, fixes them, checks the fix worked, and remembers what it learned. We're not there yet. But we're closer than you might expect.

Reviewer: ```suggestion
We built it to chase one idea: self-driving product development. A self-reinforcing system that identifies problems, fixes them, checks the fix worked, and remembers what it learned.
```

Reviewer: I don't think we need to undercut ourselves by saying we're not there yet.

- **Level 0 – No automation.** Dashboards, SQL, docs. Engineers do everything. This was most of what shipped before 2024.
- **Level 1 – Engineer assistance.** The agent answers narrow questions and writes short snippets when you point at the file. Ambiguity breaks it.
- **Level 2 – Partial automation.** The agent handles vague questions, stitches queries together, and writes whole functions. You still decide what to ask.
- **Level 3 – Conditional automation.** The agent does end-to-end work in its domain – drafts whole PRs, runs multi-step analyses, builds dashboards from a vague brief. It asks follow-up questions and proposes next steps.
- **Level 4 – High automation.** The agent understands your code, your data, and the world around it. It proposes and ships work on its own, inside guardrails you set.
- **Level 5 – Full automation.** The loop runs itself. Signals come in, work ships, evals check it, memory updates. Engineers still drive their own work – the loop just handles what doesn't need their attention.

> Here's where things sit today: [PostHog AI](/ai) is at Level 3. You prompt it in chat, it pulls its own context, and suggests its own follow-ups – but it stops at the codebase. PostHog Code picks up where it leaves off.

Reviewer: ```suggestion
```

We should cut this entire section and split it out into a separate blog post / social post at some point. It just slows the whole piece down . The "How we're defining self-driving" section" is a much better continuation of the intro. This is too "in the weeds"

## How we're defining self-driving

The *self* in self-driving isn't autonomy from the engineer. It's autonomy from instruction.

There's a lot of work that distracts engineers from building things that could actually be put on cruise control: bugs, UX issues, small features, conversion-rate tweaks. The work that drains engineering hours but rarely needs strategic input.

![Complexity pyramid](https://res.cloudinary.com/dmukukwp6/image/upload/complexity_pyramid2_324f0c5d8c.png)

This is where product data does the heavy lifting. In a typical week, PostHog customers generate 100,000+ failed queries and ~1.5 million new error tracking issues. Each one is a signal an agent could act on.

Acting on signals takes more than writing code. Claude Code, Codex, and others already do that part well. To make them self-driving, we add 5 things on top: **tools, skills, signals, memory,** and **evaluation**.

> Our [AI engineering handbook](/handbook/engineering/ai/ai-platform) goes deep on each. Here's the short version.

Reviewer: @cleo-pleurodon this a strong section. we just need to add a little more context in the top to add back in some of the context from the six levels section we're taking out.

PostHog Code has skills for the workflows we see most – instrumenting events, auditing flags, adding error tracking. Writing a skill feels like writing a doc, and [most engineers would rather ship the feature](/newsletter/agent-first-product-engineering) than document it. But for an agent, the skill *is* the feature.

### 3. Signals – when the work should happen

Tools and skills cover the *what*. Signals cover the *when*. PostHog Code sits on top of your product data, so the pattern itself is the prompt. Raw observations get grouped, enriched, and turned into concrete plans. You open a pre-framed to-do list instead of triaging a noisy inbox.

### 4. Memory – what the agent already knows

Signals say something's happening now. Memory says what happened last time. Without it, the agent [re-runs work it's already done](/blog/optimizing-agent-cost) and re-opens the same PR every Tuesday.

### 5. Evaluation – did it actually work?

> The loop doesn't close without this. [Testing AI agents](/blog/testing-ai-agents) looks nothing like testing normal software, and "it runs without erroring" is not a pass. PostHog Code schedules evals as long-running Temporal jobs, so the check runs hours or days after the PR merges. The same dashboard, funnel, experiment, or [LLM-as-a-judge eval](/blog/stop-ai-slop) that fired the signal gets re-queried. If the metric didn't move – or moved the wrong way – the agent reverts or reopens the work.

Reviewer: ```suggestion
The loop doesn't close without this. [Testing AI agents](/blog/testing-ai-agents) looks nothing like testing normal software, and "it runs without erroring" is not a pass. 

PostHog Code schedules evals as long-running Temporal jobs, so the check runs hours or days after the PR merges. The same dashboard, funnel, experiment, or [LLM-as-a-judge eval](/blog/stop-ai-slop) that fired the signal gets re-queried. If the metric didn't move – or moved the wrong way – the agent reverts or reopens the work.
```

## The product autonomy loop

Stitch those five things together, and you get a loop we like to call product autonomy:

*Collect data → cluster signals → check memory → notify workers → do work → evaluate → write back to memory.*

![Loop diagram](TODO_loop_diagram_image_url)

You can't run this reliably inside a general-purpose coding agent, because the critical signals live somewhere else. For a lot of companies, that somewhere else is PostHog.

**How this works in PostHog Code:**

> Errors, replays, and external signals flow into the signals pipeline and are clustered into signal reports. Each task that lands in your inbox is ranked by urgency, and links to relevant context and research done by background agents.

Reviewer: ```suggestion
Errors, replays, and external signals flow into the signals pipeline and are clustered into signal reports. 

Each task that lands in your inbox is ranked by urgency, and links to relevant context and research done by background agents.
```

> ![Signals](https://res.cloudinary.com/dmukukwp6/image/upload/signals_full_light_1ceb88aa2c.png)

Reviewer: we need a cleaner "less noisy" version of this screenshot with some annotation to explain what I'm looking at. In isolation it's just too much to parse at a glance.

> You pick the tasks worth working on and the model and harness that fit each one. The split-screen Command Center handles up to nine agents in parallel – our engineers call it *dopamine mode* (you'll understand why). Long jobs run in the cloud so your laptop stays free.

Reviewer: ```suggestion
You pick the tasks worth working on, and the model and harness that fit each one. 

The split-screen Command Center handles up to nine agents in parallel – our engineers call it *dopamine mode* (you'll understand why). 

Long jobs run in the cloud, so your laptop stays free.
```

![Command Center](https://res.cloudinary.com/dmukukwp6/image/upload/command_center_light_096957d499.png)

> The PostHog side is wired in by default. One-click instrumentation drops events, flags, and experiments into your code without you typing the boilerplate. The [PostHog MCP](/docs/model-context-protocol) handles impact measurement, [error debugging](/docs/error-tracking/debug-errors-mcp), and dashboard building. Plug in other MCP servers to take more actions or pull in extra context as you build.

Reviewer: ```suggestion
The PostHog side is wired in by default. One-click instrumentation drops events, flags, and experiments into your code without you typing the boilerplate. 

The [PostHog MCP](/docs/model-context-protocol) handles impact measurement, [error debugging](/docs/error-tracking/debug-errors-mcp), and dashboard building. 

Plug in other MCP servers to take more actions or pull in extra context as you build.
```

With the routine work automated, you get more room for the big stuff. Prompt your own tasks, build with full product-data context, and keep shipping new features alongside the self-driving work.

## Try PostHog Code

> Product engineers [keep telling us this is the missing piece](/newsletter/agent-first-product-engineering) – an agent that understands your codebase *and* your product. We're not fully self-driving yet. But we're getting closer every week, and you get to watch it happen from the driver's seat.

Reviewer: ```suggestion
Product engineers [keep telling us this is the missing piece](/newsletter/agent-first-product-engineering) – an agent that understands your codebase *and* your product. 

We're not fully self-driving yet, but we're getting closer every week, and you get to watch it happen from the driver's seat.
```

---
date: "2026-05-05"
title: "The self-driving product (and why we're launching PostHog Code)"
author:
  - cleo-lant
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/selfdriving_3f4c644a54.png
featuredImageType: full
category: PostHog news
tags:
  - Product updates
  - PostHog news
---

Today, [PostHog Code](/code) enters beta. 

It's a desktop app that runs coding agents on top of your product data. The obvious stuff ships itself. The tricky stuff shows up as a prioritized to-do list for you to steer.

We built it to chase one idea: self-driving product development.

## How we're defining self-driving

A self-driving product can prompt itself. 

It understands your codebase, your data, and your users. It proposes and ships work on its own, inside of guardrails you set.

The *self* in self-driving isn't autonomy from the engineer. It's autonomy from user instruction as the starting point. 

![Complexity pyramid](https://res.cloudinary.com/dmukukwp6/image/upload/complexity_pyramid_ef1350e915.png)

> There's a lot of work that distracts engineers from building things that could actually be put on cruise control: bugs, UX issues, small features, conversion-rate tweaks – things that drain engineering hours but rarely needs strategic input.

Reviewer: ```suggestion
A self-driving product puts put the 1% gains on cruise control: the bugs, UX issues, paper cuts, and conversion tweaks. The things that drain engineering hours, but rarely need strategic input. 
```

> This is where product data does the heavy lifting. In a typical week, PostHog customers generate 100,000+ failed queries and ~1.5 million new error tracking issues. Each one is a signal an agent could act on.

Reviewer: ```suggestion
In a typical week, PostHog customers generate 100,000+ failed queries and ~1.5 million new error tracking issues. Each one is a signal an agent could act on.
```

---

> date: "2026-04-20"

Reviewer 2: update date when publish

title: "The self-driving product (and why we're launching PostHog Code)"
author:
  - cleo-lant
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: >-

>   https://res.cloudinary.com/dmukukwp6/image/upload/Film_Noir_Hog_Evals_beta_announcement_blog_95685493eb.png

Reviewer 2: don't forget to update too

- **Level 0 – No automation.** Dashboards, SQL, docs. Engineers do everything. This was most of what shipped before 2024.
- **Level 1 – Engineer assistance.** The agent answers narrow questions and writes short snippets when you point at the file. Ambiguity breaks it.
- **Level 2 – Partial automation.** The agent handles vague questions, stitches queries together, and writes whole functions. You still decide what to ask.
- **Level 3 – Conditional automation.** The agent summarizes thousands of signals and drafts a PR. You review instead of originate.
- **Level 4 – High automation.** The agent understands your code, your data, and the world around it. It proposes and ships work on its own, inside guardrails you set.
- **Level 5 – Full automation.** The loop runs itself. Signals come in, work ships, evals check it, memory updates. You set strategy.

[PostHog AI](/ai) crossed into Level 3 in late 2025. The [PostHog MCP](/docs/model-context-protocol) makes Level 4 possible today. Level 5 is what we're building toward.

## Why vibe-coding tools plateau at Level 3

> A self-driving car isn't a vehicle that goes where you point it (that's cruise control). A self-driving product isn't just a really good coding agent. To reach Level 5, it has to sense what's happening, decide what to do, do it, check if it worked, and adjust next time — all without being asked. The *self* in self-driving isn't autonomy from the engineer. It's autonomy from instruction.

Reviewer 2: This is a standout point I think
```suggestion
A self-driving car isn't a vehicle that goes where you point it (that's cruise control). A self-driving product isn't just a really good coding agent. To reach Level 5, it has to sense what's happening, decide what to do, do it, check if it worked, and adjust next time — all without being asked. 

The *self* in self-driving isn't autonomy from the engineer. It's autonomy from instruction.
```

> Vibecoding tools aren't aiming for that. They're aiming for speed per instruction – and they've gotten *very* good at it. Point Lovable, Replit or [v0](/blog/posthog-v0-integration) at a goal and they'll ship a working app before lunch. That's a legitimate Level 3 experience for *building* software, and it's why [AI has quietly eaten the no-code category](/blog/ai-is-killing-no-code-experiments).

Reviewer 2: ```suggestion
Vibe-coding tools aren't aiming for that. They're aiming for speed per instruction – and they've gotten *very* good at it. Point Lovable, Replit or [v0](/blog/posthog-v0-integration) at a goal and they'll ship a working app before lunch. That's a legitimate Level 3 experience for *building* software, and it's why [AI has quietly eaten the no-code category](/blog/ai-is-killing-no-code-experiments).
```

![On Wednesdays we do trace reviews](https://res.cloudinary.com/dmukukwp6/image/upload/aps0un_8b5268c19e.jpg)

> The problem is what happens after lunch. In a typical week, PostHog customers generate 100,000+ failed queries, 1.2 to 2.6 million new error-tracking issues, and 12 to 47 million raw exceptions. Each one is a signal an agent could act on. None of it is visible to a tool whose worldview ends at the last git push.

Reviewer 2: Let's just simplify and give one number
```suggestion
The problem is what happens after lunch. In a typical week, PostHog customers generate 100,000+ failed queries, 2.6 million new error tracking issues, and 47 million raw exceptions. Each one is a signal an agent could act on. None of it is visible to a tool whose worldview ends at the last git push.
```

To move from cruise control to self-driving, an agent needs five things: **tools, skills, signals, memory,** and **evaluation**. Our [AI engineering handbook](/handbook/engineering/ai/ai-platform) goes deep on each. Here's the short version.

### 1. Tools – what the agent can do

> Tools are small, specific actions the agent can take. At PostHog we treat them as [atomic capabilities](/handbook/engineering/ai/implementation#mcp-tools-vs-skills) — things like `Create_insight` or `read_taxonomy`. That last one does a lot of heavy lifting. It lets the agent check what events and properties actually exist before it writes a query or an instrumentation PR.

Reviewer 2: Do you have one more tool we could include, it makes "that last one" read better. Otherwise, we could say "the latter"

```suggestion
Tools are small, specific actions the agent can take. At PostHog we treat them as [atomic capabilities](/handbook/engineering/ai/implementation#mcp-tools-vs-skills) — things like `create_insight` or `read_taxonomy`. The latter does a lot of heavy lifting. It lets the agent check what events and properties actually exist before it writes a query or an instrumentation PR.
```

### 2. Skills – how to get the job done

> If tools are the fork and knife, skills are the recipe. A [skill](/handbook/engineering/ai/writing-skills) ties tools, docs, and rules into a playbook. You can see the gap in our own data: `docs-search` is the most-called tool on PostHog's MCP server, at ~28K calls a month. That's every coding agent asking our docs for directions. PostHog Code has skills for the workflows we see most — instrumenting events, auditing flags, adding error tracking. Writing a skill feels like writing a doc, and [most engineers would rather ship the feature](/newsletter/agent-first-product-engineering) than document it. But for an agent, the skill *is* the feature.

Reviewer 2: ```suggestion
If tools are the fork and knife, skills are the recipe. A [skill](/handbook/engineering/ai/writing-skills) ties tools, docs, and rules into a playbook. You can see the gap in our own data: `docs-search` is the most-called tool on PostHog's MCP server, at ~28K calls a month.

PostHog Code has skills for the workflows we see most – instrumenting events, auditing flags, adding error tracking. Writing a skill feels like writing a doc, and [most engineers would rather ship the feature](/newsletter/agent-first-product-engineering) than document it. But for an agent, the skill *is* the feature.
```

### 3. Signals – when the work should happen

Tools and skills cover the *what*. Signals cover the *when*. PostHog Code sits on top of your product data, so the pattern itself is the prompt. Raw observations get grouped, enriched, and turned into concrete plans. You open a pre-framed to-do list instead of triaging a noisy inbox.

### 4. Memory – what the agent already knows

Signals say something's happening now. Memory says what happened last time. Without it, the agent [re-runs work it's already done](/blog/optimizing-agent-cost) and re-opens the same PR every Tuesday.

### 5. Evaluation – did it actually work?

> The loop doesn't close without this. [Testing AI agents](/blog/testing-ai-agents) looks nothing like testing normal software, and "it runs without erroring" is not a pass. PostHog Code schedules evals as long-running Temporal jobs, so the check runs hours or days after the PR merges. The same dashboard, funnel, experiment, or [LLM-as-a-judge eval](/blog/stop-ai-slop) that fired the signal gets re-queried. If the metric didn't move — or moved the wrong way — the agent reverts or reopens the work.

Reviewer 2: ```suggestion
The loop doesn't close without this. [Testing AI agents](/blog/testing-ai-agents) looks nothing like testing normal software, and "it runs without erroring" is not a pass. PostHog Code schedules evals as long-running Temporal jobs, so the check runs hours or days after the PR merges. The same dashboard, funnel, experiment, or [LLM-as-a-judge eval](/blog/stop-ai-slop) that fired the signal gets re-queried. If the metric didn't move — or moved the wrong way – the agent reverts or reopens the work.
```

## The product autonomy loop

Stitch the five together and you get a loop:

*Collect data → cluster signals → check memory → notify workers → do work → evaluate → write back to memory.*

That's what a self-driving product runs on. You can't do this reliably inside a general-purpose coding agent, because the critical signals live somewhere else. For a lot of companies, that somewhere else is PostHog.

![we added skills, memory and a flux capacitor](https://res.cloudinary.com/dmukukwp6/image/upload/no_roadmaps_d355fa4d19.jpg)

> ## Why we built PostHog Code (and what it does for you)

Reviewer 2: nit but this section doesn't cover so much the "why" we mostly cover that before. don't feel strongly if you like the other title.
```suggestion
## What PostHog Code does for you
```

We've spent years building the stack this loop needs:

- **Data** – [errors](/docs/error-tracking/start-here), [session replays](/docs/session-replay/start-here), [LLM traces](/docs/llm-analytics/start-here), [experiments](/docs/experiments/start-here), [feature flags](/docs/feature-flags/start-here), revenue. What real users are actually doing.
- **Warehouse** – product events plus anything you sync from Stripe, Salesforce, Snowflake, BigQuery, or flat files. ClickHouse today, DuckDB soon.
- **Orchestration** – Temporal for long-running jobs that outlive any one agent session.
- **Sandboxes** – cloud agents that run code safely outside of production.

> - **Devtools** – the things product engineers actually reach for.

Reviewer 2: Maybe we include an MCP mention here?

Author: very good idea

PostHog Code is the interface that ties it all together. It does two things.

**It ships PRs for the obvious stuff.** Something breaks, the agent pulls in context, writes the fix, opens a PR, and schedules an evaluation. If the metric moves the wrong way, it reverts itself.

**It frames the judgement calls for you.** Ten support tickets about a confusing checkout? PostHog Code checks the replays, quantifies the drop-off, and drafts a PR. You iterate, kill, or deploy.

Under the hood, it's a multi-harness, multi-agent orchestrator. You can run Claude Code and Codex side by side, each on its own task, with their own model and settings. A split-screen Command Center lets you run up to nine agents in parallel – our engineers call it *dopamine mode* (you'll understand why).

Git and GitHub are native. Branches, commits, PRs, and review all happen inside the app, with AI-written branch names, commit messages, and PR descriptions. Long jobs run in the cloud so your laptop stays free; local worktrees handle the rest.

> The PostHog side has built-in skills for instrumenting events, flags, and experiments, plus the PostHog MCP for measuring impact, debugging errors, and building dashboards. Plug in your own MCP servers (Attio, Canva, Atlassian, Notion, Monday) and pull context from Linear, Zendesk, and GitHub into the same pipeline as your product data.

Reviewer 2: For the sake of brevity, could we blend these two paragraphs into the "stack" list at the top of this section, they feel pretty close to duplicating what's said there.

Reviewer 2: Maybe the Git stuff needs to stay

Reviewer 2: Yeah, I think looking at the preview, this section is too long.

> You get a prioritized to-do list instead of a noisy one, and automatic fixes for the stuff that doesn't need your attention.

Author: demo video will replace this line

> Product engineers [keep telling us this is the missing piece](/newsletter/agent-first-product-engineering) — an agent that understands your codebase *and* your product. We're not at Level 5 yet. But the gap closes every week, and you get to watch it happen from the driver's seat.

Reviewer 2: ```suggestion
Product engineers [keep telling us this is the missing piece](/newsletter/agent-first-product-engineering) – an agent that understands your codebase *and* your product. We're not at Level 5 yet. But the gap closes every week, and you get to watch it happen from the driver's seat.
```

Critically, it does all of that continuously, without being asked. The *self* in self-driving isn't autonomy from the engineer. It's autonomy from instruction.

Vibecoding tools aren't aiming for autonomy from instruction. They're aiming for speed per instruction – and they've gotten *very* good at it. Apps like Lovable, Replit and [v0](/blog/posthog-v0-integration) have proven you can get surprisingly far by equipping agents with tools and skills: plug in the right MCP servers, point the agent at a goal, ship a working app before lunch. That's a legitimate Level 3 experience for *building* software, and it's why [AI has quietly eaten the no-code category](/blog/ai-is-killing-no-code-experiments).

![On Wednesdays we do trace reviews](https://res.cloudinary.com/dmukukwp6/image/upload/aps0un_8b5268c19e.jpg)

> The problem is what happens after lunch. In a typical week, PostHog customers generate 100,000+ failed queries, 1.2 to 2.6 million new error-tracking issues, and 12 to 47 million raw exceptions. Each one is a candidate signal for an agent to act on – and none of it is visible to a tool whose worldview ends at the last git push.The signals the agent needs next don't live in the codebase. In a typical week, PostHog customers generate 100,000+ failed queries, 1.2 to 2.6 million new error-tracking issues, and 12 to 47 million raw exceptions. Each one is a candidate signal for an agent to act on – and none of it is visible to a tool whose worldview ends at the last git push.

Reviewer 3: ```suggestion
The problem is what happens after lunch. In a typical week, PostHog customers generate 100,000+ failed queries, 1.2 to 2.6 million new error-tracking issues, and 12 to 47 million raw exceptions. Each one is a candidate signal for an agent to act on – and none of it is visible to a tool whose worldview ends at the last git push. The signals the agent needs next don't live in the codebase.
```

Proactive work is what happens when you prompt yourself – when you shift out of maintenance mode and decide what to build next. It needs context, judgement, and that other word everyone won't stop saying: [taste](/newsletter/good-taste-great-products).

There's a third category in between, and it's the one getting interesting: proactive work a computer could do if it had the tools, context, and permissions. That's where self-driving products live, and the lane we're launching a new product into.

Today, [PostHog Code](/code) enters beta. It's a desktop app for agentic engineering that automates the reactive parts of product building and frames the proactive work for you to steer.

## Borrowing a taxonomy from the car people

The Society of Automotive Engineers has a framework for self-driving. Six levels of autonomy, from Level 0 (you do everything) to Level 5 (the car does everything, everywhere, in any condition). It's a good mental model because it forces specificity about where the machine takes over and where a human is still on the hook.

> It maps surprisingly cleanly onto building products with AI.

Reviewer 4: nit - you imply you're going to, but you don't use the _actual_ framework below ie 'actionable product guidance' i assume is just 'actionable guidance' in their framework

Author: ah that's a good nit. The exact framework is 
SAE Level 0: No automation
SAE Level 1: Driver assistance
SAE Level 2: Partial automation
SAE Level 3: Conditional automation
SAE Level 4: High automation
SAE Level 5: Full automation

I will update the names to map more clearly.

Which brings us back to Level 5 – and the question I ducked earlier.

![we added, skills, memory and a flux capacitor](https://res.cloudinary.com/dmukukwp6/image/upload/no_roadmaps_d355fa4d19.jpg)

## Why we built PostHog Code (and what that means for you)

We've been sitting on a weirdly complete set of ingredients for this loop:

- **Data** – years of it. [Errors](/docs/error-tracking/start-here), [session replays](/docs/session-replay/start-here), [LLM traces](/docs/llm-analytics/start-here), [experiments](/docs/experiments/start-here), [feature flags](/docs/feature-flags/start-here), revenue. The event stream of what real users are actually doing.
- **Warehouse** – your full data stack in one place: product events, application data, and anything you sync from Stripe, Salesforce, Snowflake, BigQuery, or flat files. Runs on ClickHouse today, DuckDB in the near future.
- **Orchestration** – Temporal for long-running jobs that outlive any single agent session, so evaluation runs hours after execution without anyone holding the process open.

> - **Sandboxes** – cloud agents for running code safely, without blowing up production.

Reviewer 4: sandboxes were introduced very recently just for this purpose so strictly speaking weren't something we happened to be sitting on

---
date: "2026-04-20"
title: "The self-driving product (and why we're launching PostHog Code)"
author:
  - cleo-lant
rootPage: /blog
sidebar: Blog
showTitle: true
hideAnchor: true
featuredImage: >-
  https://res.cloudinary.com/dmukukwp6/image/upload/Film_Noir_Hog_Evals_beta_announcement_blog_95685493eb.png
featuredImageType: full
featuredImageCaption: The self-driving product is coming. PostHog Code is the workbench we're building it in.
category: PostHog news
tags:
  - Product updates
  - PostHog news
seo:
  metaTitle: "The self-driving product (and why we're launching PostHog Code)"
  metaDescription: "Six levels of product autonomy, the five things every agent actually needs, and why we're shipping PostHog Code into beta."
---

import PostHogCodeWaitlist from 'components/PostHogCodeWaitlist'

Product work splits two ways: reactive and proactive.

> Reactive work is mechanical. Fix the typo. Add the index. Flip the flag. The data tells you what's wrong and the scope tells you what to fix. A computer can do this better than you. Full stop.

Reviewer 2: The first thought that comes to mind about "reactive" work is an incident which a computer can't seemingly do better than humans yet, they don't know what data to care about? 

Customer issues also seem like reactive work, but they require interpreting what they want, fitting it into an existing product, and prioritizing. 

Maybe this is me being too bearish on AI but it seems a bit hyperbolic at the moment.

Reviewer 2: Reading further on, it's more like product work splits three ways and these two are the "in-between" third one but I get stuck on the second sentence because of the way this piece starts out.

Author: Easy fix. Reads much stronger now

Proactive work is what happens when you prompt yourself – when you shift out of maintenance mode and decide what to build next. It needs context, judgement, and that other word everyone won't stop saying: [taste](/newsletter/good-taste-great-products).

There's a third category in between, and it's the one getting interesting: proactive work a computer could do if it had the tools, context, and permissions. That's where self-driving products live, and the lane we're launching a new product into.

> Today, [PostHog Code](/code) enters beta. It's a desktop app for agentic engineering that automates the reactive parts of product building and frames the proactive work for you to steer. It's also the first public piece of a much bigger bet (more on that later).

Reviewer 2: Let's just really focus on PostHog Code for now. 

```suggestion
Today, [PostHog Code](/code) enters beta. It's a desktop app for agentic engineering that automates the reactive parts of product building and frames the proactive work for you to steer.
```

> If you've been following [our AI work](/blog/8-learnings-from-1-year-of-agents-posthog-ai), you probably saw this coming. If not, you might reasonably wonder why we think we can win in a space that has Cursor, Claude Code, Codex, and a new entrant every other Tuesday. This post is the map, how we're thinking about it, and why I think you should give our version of agentic engineering a shot.

Reviewer 2: Do we want to anchor ourselves in Cursor, Claude Code, etc. or become a "special different category"?

Also, just based on social posts, I'm pretty sure basically no one saw this coming.

Reviewer 2: I might just drop this paragraph to simplify and get to the meat faster.

I think it doesn't add so much new information. We don't need to tell people "this is how we are thinking about this", it is implied by writing about it.

## Borrowing a taxonomy from the car people

> The Society of Automotive Engineers has had a framework for self-driving since 2014. Six levels of autonomy, from Level 0 (you do everything) to Level 5 (the car does everything, everywhere, in any condition). It's a good mental model because it forces specificity about where the machine takes over and where a human is still on the hook.

Reviewer 2: ```suggestion
The Society of Automotive Engineers has a framework for self-driving. Six levels of autonomy, from Level 0 (you do everything) to Level 5 (the car does everything, everywhere, in any condition). It's a good mental model because it forces specificity about where the machine takes over and where a human is still on the hook.
```

> It maps surprisingly cleanly onto building products with AI.

Reviewer 2: The whole list below focuses on "data analysis" basically, but the "writing code" is arguably the bigger part of "building products"

There is also an interpretation by engineers bit that might be missing. As autonomy increases, engineers don't need to know which questions to ask as much or how to interpret the data.

**Level 0 – No automation.** Analytics without AI. Dashboards, filters, SQL. Everything you do, you do yourself through an interface. Really good docs are your best bet for troubleshooting. This is most of what shipped before 2024.

**Level 1 – Basic data analysis.** The agent answers specific, well-formed questions like *"How many users are on the pro plan?"* or *"What's our conversion rate?"* It falls apart when the question is ambiguous, or the query engine can't express it. This was the first version of [PostHog AI](/ai).

**Level 2 – Versatile data analysis.** The agent handles ambiguity and stitches queries together. It answers vague, loaded questions like *"Which features do users engage with most?"* PostHog AI [reached this state in 2025](/blog/8-learnings-from-1-year-of-agents-posthog-ai), once reasoning models matured and we learned how to harness them.

> **Level 3 – Actionable product guidance.** The agent compresses massive amounts of context and responds not just to the query, but suggests meaningful follow-up actions. Late 2025 is when PostHog AI started to feel like the future instead of a fancy search bar – it could summarize thousands of session recordings *and* give excellent suggestions for what to do next.

Reviewer 2: ```suggestion
**Level 3 – Actionable product guidance.** The agent compresses massive amounts of context and responds not just to the query, but suggests meaningful follow-up actions. Late 2025 is when PostHog AI started to feel like the future instead of a fancy search bar – it could summarize thousands of session recordings *and* give suggestions for what to do next.
```

**Level 4 – Proactive product maintenance.** The agent understands your codebase, your product data, and the surrounding context – support tickets, internal comms, industry news. It proposes work without being asked and has permissions to ship autonomously. The [PostHog MCP](/docs/model-context-protocol) enables a working prototype of this. You can hand product data to your agent and automate scoped work like bug fixes.

> **Level 5 – Self-driving development.** An agentic system that decides, orchestrates, and evaluates itself. You might prompt work and weigh in on decisions, but it's largely self-sufficient. *"Just build me an app that generates $1M MRR, bro"* is a plausible starting prompt. That's not why you became a product engineer, though – so what does self-driving actually mean for the human in the loop?

Reviewer 2: is it a plausible prompt?

## Code generation ≠ building software

Allow me to dodge that question for a second and make a clear distinction. A self-driving car isn't just a vehicle that gets itself where you point it (that's cruise control and lane assist).

An autonomous vehicle must know where it is, what's around it, what those things might do next, and what it will do in response. Once it executes a decision, it must evaluate whether the execution matched the plan, and adjust the next decision based on what it learned.

Critically, it does all of that continuously, without being asked. The *self* in self-driving isn't autonomy from the passenger. It's autonomy from instruction.

> Apply the same variables to software and the framework looks similar:

Reviewer 2: To be honest, this whole list feels quite "heavy" intellectually like we just did a whole list on the self-driving product and now we are doing another one basically on what level 5 would need?

I would love to read about PostHog Code by now, it feels a bit lost. I'm craving some more about PostHog Code at this point rather than another framework!

If not, maybe "Why vibe-coding tools plateau at Level 3" would be better to put ahead of this. 

Maybe we could even remove this section and blend this list into the "an agentic engineering tool needs five things: **tools, skills, signals, memory,** and **evaluation**." in that section because it feels like they are related.

Author: Tough but fair. removed this whole section and will use it for a social post or something

> - **Localization** – it knows what state it's in. Which users are doing what, which features are healthy, which experiments are running, what shipped yesterday.

Reviewer 2: What is "it?" The agent? The self-driving product?

Maybe we need to change the word "software" in "apply the same variables to software"

- **Control** – it has what it needs to act without asking. Skills to implement feature flags, permission to merge into master.
- **Feedback** – it checks whether the execution worked. Did engagement go up? Did latency drop? Did adding a tooltip have a meaningful impact?
- **Learning** – most importantly, it updates the next decision based on that feedback. If the fix doesn't work, it tries something else. If the experiment was conclusive, it rolls out the best variant. If you told it not to use em dashes, it actually listens to you.

## Why vibe-coding tools plateau at Level 3

Lovable, Replit, Bolt, v0, and the rest of the vibe-coding cohort have proven you can get surprisingly far by equipping agents with tools and skills. Plug in the right MCP servers, point the agent at a goal, ship a working app before lunch. That's a legitimate Level 3 experience for *building* software, and it's why [AI has quietly eaten the no-code category](/blog/ai-is-killing-no-code-experiments).

![On Wednesdays we do trace reviews](https://res.cloudinary.com/dmukukwp6/image/upload/aps0un_8b5268c19e.jpg)

> Lovable itself is already one of the top clients hitting PostHog's MCP server, and [PostHog is a first-class integration in Vercel's v0](/blog/posthog-v0-integration). The vibe-coding cohort isn't the competition here – they're customers reaching for product data to do the Level 3 job better. The ceiling they can't cross is the one that sits above that: the proactive work that starts *after* the app ships.

Reviewer 2: Do we have stats on "top clients hitting PostHog's MCP server" ?

The Lovable being a top client seems not true unfortunately: https://us.posthog.com/project/2/home?mode=ai&chat=0c9700ce-44f4-42f3-9953-3d98cf9c0a4a (maybe I am misunderstanding).

> In a typical week, PostHog customers generate 100,000+ failed queries, 1.2 to 2.6 million new error-tracking issues, and 12 to 47 million raw exceptions. Each one is a candidate signal for an agent to act on – and none of it is visible to a tool whose worldview ends at the last git push.

Reviewer 2: Great one!

> To shift out of cruise control and into self-driving, an agentic engineering tool needs five things: **tools, skills, signals, memory,** and **evaluation**. Our [AI engineering handbook](/handbook/engineering/ai/ai-platform) goes deep on each one. Here's the quick version.

Reviewer 2: This isn't that quick. Could we:

1. Make each of them 1/2 paragraphs
2. Focus a lot more on all of these in the context of PostHog Code?

### 1. Tools – atomic capabilities

Tools are agent-friendly descriptions of what your product can do. At PostHog, we treat them as [atomic capabilities](/handbook/engineering/ai/implementation#mcp-tools-vs-skills) – CRUD operations and simple actions that answer *"what can I do?"* Create an insight. Execute a query. Read the taxonomy. Flip a flag. Agents are good at composing simple tools; they're bad at guessing which composite tool you secretly wish you'd built.

One atomic tool in particular is doing disproportionate work at PostHog: `read_taxonomy`. It lets the agent check what events and properties actually exist before it writes a query. No `read_taxonomy`, no [hallucination protection](/handbook/engineering/ai/architecture).

> MCP tool calls against our server have grown ~21× in five months – 60K in October 2025 to 1.26M in March 2026, across 70+ clients including Claude Code, Cursor, VS Code, Codex, Lovable, Windsurf, Copilot, and Gemini CLI. If you haven't shipped an MCP server or a machine-readable API surface, you're invisible in this stack. Tools are table stakes.

Reviewer 2: Don't know if this is super relevant here, we have already sort of mentioned in Lovable MCP client calls this idea.

### 3. Signals – when work should happen

Tools and skills guide the *what*. Signals determine the *when*.

An exception spike, a checkout funnel drop, a replay of someone rage-clicking the same button three times, a cluster of support tickets that sound suspiciously alike – these are signals. Most products have tools. A few have skills. Almost nobody has signals. The [PostHog AI architecture](/handbook/engineering/ai/ai-platform#vision-product-autonomy) treats signals as the entry point to the autonomy loop: raw observations get deduplicated, contextualized, and turned into concrete plans an agent can execute.

### 4. Memory – what the agent already knows

Signals tell the agent something is happening *now*. Memory tells it what happened last time, and the time before that, and whether what it's about to try has already failed.

Without memory, the agent is a goldfish with git access. It re-opens the same PR every Tuesday because the same signal fires every Tuesday. Memory also keeps costs sane – we [caught our own AI agent embezzling tokens](/blog/optimizing-agent-cost) because it kept re-running work it had already done. Durable state prevents that.

> Memory isn't a context window. It's a persistent model of the product's own history: every PR the agent has opened, linked to the signal that triggered it, linked to the metric that confirmed or denied the fix, queryable by the next agent run. PostHog's [single-loop architecture](/handbook/engineering/ai/architecture) is our version of this – the agent maintains full context across decisions instead of handing off to amnesiac sub-agents. Nearly every other coding agent on the market still skips durable memory in favor of a context window that resets the moment you start a new session.

Reviewer 2: Needs tightening

### 5. Evaluation – did the thing actually work?

The loop doesn't close without this. An agent can ship a PR. A self-driving product ships a PR and then checks whether the thing it was trying to fix actually got fixed – against the same signal that triggered the work in the first place.

Evals are the hard part nobody was trained for. [Testing AI agents](/blog/testing-ai-agents) looks nothing like testing traditional software, and "it runs without erroring" is not a pass condition. We've written about running [LLM-as-a-judge evals](/blog/stop-ai-slop) against your own agent output for exactly this reason. At PostHog, evaluation is the last leg of the loop: a long-running job that re-queries the same dashboard, funnel, or experiment that fired the signal, and decides whether the PR moved the needle or needs a revert.

> ## The loop

Reviewer 2: ```suggestion
## The product autonomy loop
```

Stitch the five together and a pattern falls out:

*Collect data → cluster signals → check memory → notify workers → do work → evaluate → write back to memory.*

That's the product autonomy loop. Tools, skills, and signals get you to Level 4. Memory and evaluation are what Level 5 is built out of. You can't reliably do this inside a general-purpose coding agent, because the signals, memory, and evaluation data all live somewhere else. For a lot of companies, that somewhere else is PostHog.

Which brings us back to Level 5 – and the question I ducked earlier.

> ![we added, skills, memory and a flux capacitor](https://res.cloudinary.com/dmukukwp6/image/upload/no_roadmaps_d355fa4d19.jpg)

Reviewer 2: good one

## Why we built PostHog Code (and what that means for you)

We've been sitting on a weirdly complete set of ingredients for this loop:

- **Data** – years of it. [Errors](/docs/error-tracking/start-here), [session replays](/docs/session-replay/start-here), [LLM traces](/docs/llm-analytics/start-here), [experiments](/docs/experiments/start-here), [feature flags](/docs/feature-flags/start-here), revenue. The event stream of what real users are actually doing.

> - **Warehouse** – ClickHouse for storing and querying events and agent decision history in the same place.

Reviewer 2: Warehouse should include data stack which means data from basically anywhere. This happens through ClickHouse now but will be DuckDB in the near future.

**Second, it frames bounded decisions so you can actually decide.** Ten support tickets about a confusing checkout flow? The agent checks the replays, quantifies the drop-off, checks memory for whether this was already discussed in a previous run, and drafts the options. You still make the call – redesign, kill, or ignore. The legwork is already done.

The same pattern plays out for the internal agents we've already built ourselves, like [the one that writes Andy's standup notes for him](/blog/standup-bot-revenge). The difference with PostHog Code is that you don't have to be an AI-pilled engineer to wire one up.

Practically, that means a prioritized to-do list instead of a noisy one, and automatic code changes for stuff that doesn't need your energy. You can switch between Claude Code and Codex from a single interface, plug in your MCP servers (Attio, Canva, Atlassian, Notion, Monday), and pull context from Linear, Zendesk, and GitHub into the same pipeline as your product data.

> ## The TL;DR (and CTA)

Reviewer 2: I think it's a bit weird to have a TL;DR at the end. 
```suggestion
## Test drive a self-driving product
```

Our long-term vision is Level 5 self-driving. We're not *quite* there yet. The current version of PostHog Code is more of a workbench. It's useful on its own, and – more importantly – it's ahead of general-purpose agents on the one axis we care about: **product context comes pre-loaded.**

> You don't have to tab-switch, copy-paste, or upload a screenshot of your own product into the agent chat like a lunatic. PostHog Code runs in the background, watches your data, fixes what's fixable, and frames what needs a judgement call. Product engineers [keep telling us this is the missing piece](/newsletter/agent-first-product-engineering) – an agent that sees what real users do, not just what your IDE sees.

Reviewer 2: Rather than "copy paste", maybe more realistically is making dozens of MCP calls, wasting a bunch of tokens on MCP calls, filling context window with useless context, having to sign up and manage a bunch of tools, or just not having that valuable context at all.
